Coordinate system improvements

charlesgregory opened the pull request:
Fixes #81, #82, #85.
-  shortform names are no longer functions they are just aliases
   - allows this `ZBHO val = ZBHO(1, 0);`
   - `ChromCoordinates` shortcuts are of form: `ChromZBHO`
- added functions for calculating intersection and union of two `Coordinates`.
- Can now use shortform for conversion `OBC newVal = val.to!OBC;`

jblachly reviewed (commented):
Looks great overall. address my comments and we can merge.

For `ChromCoordinates` could you use htslib region parsing code for a potential speed boost (and bug freeness?) Or too hard to use?

charlesgregory commented on source/dhtslib/coordinates.d:
```diff
@@ -67,6 +68,7 @@ struct Coordinate(Basis bs)
 
     invariant
     {
+        assert(pos >= 0);
```
`Coordinates` should inherently have this check as it uses two `Coordinate`s as the start and end. Though we can add an explicit check if you would like.

charlesgregory commented on source/dhtslib/coordinates.d:
```diff
@@ -256,6 +271,99 @@ struct Coordinates(CoordSystem cs)
+    auto offset(T)(T off)
+    if(isIntegral!T)
+    {
+        return Coordinates!(cs)(cast(long)(this.start + off), cast(long)(this.end + off));
```
I had intended for most operations with `Coordinates` to return a copy. I tend to prefer this but it can easily be changed. Honestly my "copy" logic only makes sense if we treat/make all coordinates as immutable.

jblachly replied:
Offset is both a verb (implying mutable) and a noun (implying a copy. Could we give it a slightly different name?  I won't be hard line on this point, just wondering. 

`withOffset`,  `offseted` (which is in fact a word ;) ), not sure what else. may be more trouble than its worth

charlesgregory commented on source/dhtslib/coordinates.d:
```diff
@@ -256,6 +271,99 @@ struct Coordinates(CoordSystem cs)
+    auto intersectImpl(Coordinates!cs other)
+    {
+        if(!this.isOverlap(other)){
+            return Coordinates!(cs).init;
```
Probably not. I wasn't sure what is normal for set logic, throw error or return an empty interval.

jblachly replied:
We'll try to make this all `@nothrow`. One day we should make all return values Sum Types and return Ok<T>, Error ala Rust ;)

charlesgregory commented on source/dhtslib/coordinates.d:
```diff
@@ -256,6 +271,99 @@ struct Coordinates(CoordSystem cs)
+    auto unionImpl(Coordinates!cs other)
+    {
+        if(!this.isOverlap(other)){
+            return Coordinates!(cs).init;
```
So yes a true union would be an array of both intervals. I figured that wouldn't be preferred. I wasn't sure what is normal for set logic, throw error or return an empty interval.

jblachly replied:
Given these constraints, it may not be reasonable to implement union on ranges. An alternative interpretation of union could be the range `[min(A,B), max(A,B)]` -- what do youy think?

I definitely agree with intersection operation though

jblachly replied:
There's actually no reason we can't return an array of results, now that I think of it. However this invokes the GC if we use a dynamic array. How free of the GC  are we in the rest of the codebase?

charlesgregory replied:
Should be pretty GC free. No usage of arrays, all types can live on the stack with the exception of the `string` in `ChromCoordinates`.

jblachly replied:
> Given these constraints, it may not be reasonable to implement union on ranges. An alternative interpretation of union could be the range `[min(A,B), max(A,B)]` -- what do youy think?

?

charlesgregory commented on source/dhtslib/coordinates.d:
```diff
@@ -256,6 +271,99 @@ struct Coordinates(CoordSystem cs)
+
+    auto getMinStart(Coordinates!cs other)
+    {
+        return this.start < other.start ? this.start : other.start;
```
Ah see the member function is not technically polymorphic for the parent struct as its coordinate system determined by the coordinate system of the struct. 
```
auto c0 =  ZBHO(0, 2); // [0, 2)
auto c1 = ZBHO(1, 3);   // [1, 3)

assert((c0 & c1) == ZBHO(1,2)); // [1, 2)

auto c2 = OBC(2, 3); // [2, 3]
assert(c2.to!ZBHO == c1);

// doesn't compile as c2 and c0 aren't the same type
assert((c0 & c2) == OBC(2,2)); 
```

jblachly replied:
Ah got it, `self` and `other` are always of the same type.

charlesgregory commented on source/dhtslib/coordinates.d:
```diff
@@ -304,6 +412,30 @@ struct ChromCoordinates(CoordSystem cs)
     }
 
+    /// Convert coordinate to another based system using shortcuts
+    auto to(T: ZBHO)()
```
They are duplicated for `ChromCoordinates` I could instead alias the `Coordinates` functions. Up to you.

jblachly replied:
Overall I'm still not sure how much I love the existence of `ChromCoordinates`. Lots of duplication when I wonder if some other simpler method of composition might exist.

For these functions, if they are identical I would just alias them.

charlesgregory replied:
No I agree. I am not sold on `ChromCoordinates`. It was a solution at the time to adapt a `string` "chr:1-2" to a set of coordinates and a string for the chromosome. Though a lot of the reader overloads can take either a `ChromCoordinates` or a `string` and a `Coordinates`. I am open to suggestions.

jblachly replied:
@charlesgregory Leave it for now and open an issue to discusss

charlesgregory commented:
> For ChromCoordinates could you use htslib region parsing code for a potential speed boost (and bug freeness?) Or too hard to use?

Potentially, I will look into htslib and see if that function is exported and how it could be used. Probably more thoroughly tested so its worth a look. Honestly, I need to overload the region structs in `dhtslib.sam.reader` we did a while ago.

jblachly commented on source/dhtslib/coordinates.d:
```diff
@@ -256,6 +271,99 @@ struct Coordinates(CoordSystem cs)
+            );
+    }
+
+    auto isOverlap(Coordinates!cs other)
```
*way

jblachly reviewed (changes requested):
In addition to prior comment on htslib's string parsing which i'd like you to answer, I am also wondering about the use of `Coordinate` and `Coordinates` which seems confusing inasmuch as their names are almost identical.  It may be too late int he game to change the latter, but what about `Interval` or `Range` instead?

charlesgregory commented:
> In addition to prior comment on htslib's string parsing which i'd like you to answer, I am also wondering about the use of Coordinate and Coordinates which seems confusing inasmuch as their names are almost identical. It may be too late int he game to change the latter, but what about Interval or Range instead?

The nice thing about version control is any prior implementations we are already using will still work. I am open to changes as long as we don't change the API too often. I would rather we have the best implementation before the paper is out as more people start looking at the code afterwards. After we are done, we can just bump the version to 0.13 or even move to version 1.0.

jblachly commented:
> > For ChromCoordinates could you use htslib region parsing code for a potential speed boost (and bug freeness?) Or too hard to use?
> 
> Potentially, I will look into htslib and see if that function is exported and how it could be used. Probably more thoroughly tested so its worth a look. Honestly, I need to overload the region structs in `dhtslib.sam.reader` we did a while ago.

Great news:

https://github.com/samtools/htslib/blob/43c127ec10d5e38c9f5b5ec3679c3745148f7c61/htslib/hts.h#L1138-L1202

charlesgregory commented:
Latest commit changes `Coordinates` to `Interval` and removes `ChromCoordinates`.

jblachly commented on source/dhtslib/coordinates.d:
```diff
@@ -370,105 +377,32 @@ struct Coordinates(CoordSystem cs)
-    auto to(T: OBC)()
-    {
-        return this.to!(CoordSystem.obc);
+auto getIntervalFromString(string region){
```
please document return type

jblachly reviewed (changes requested):
document the new function and we'll mefrge

jblachly commented on source/dhtslib/coordinates.d:
```diff
@@ -509,3 +532,22 @@ debug(dhtslib_unittest) unittest
+    hts_log_info(__FUNCTION__, "Loading test file");
+    auto sam = SAMFile(buildPath(dirName(dirName(dirName(dirName(__FILE__)))),"htslib","test","auxf#values.sam"), 0);
+    
+    auto reg = getIntervalFromString("sheila:1-10",sam.header);
```
This fails to compile

charlesgregory commented:
I have addressed the remaining concerns in the last two commits. Since we are likely to change this implementation further anyway, I am going to accept this pull request so I can begin further work.
